packages/beacon-node/src/chain/bls/multithread.ts

```diff
@@ -348,34 +372,64 @@ export class BlsMultiThreadWorkerPool implements IBlsVerifier {
     // and to prevent large stacks since runJob may be called recursively.
     else {
       if (job.opts.priority) {
-        this.jobs.unshift(job);
+        this.jobsForNextRun.unshift(job);
       } else {
-        this.jobs.push(job);
+        this.jobsForNextRun.push(job);
       }
       setTimeout(this.runJob, 0);
     }
   }
 
   /**
-   * Potentially submit jobs to an idle worker, only if there's a worker and jobs
+   * Add all buffered jobs to the job queue and potentially run them immediately
    */
+  private runBufferedJobs = (): void => {
+    if (this.buffer) {
+      for (const job of this.buffer.jobs) {
+        this.jobsForNextRun.push(job);
+      }
+      for (const job of this.buffer.prioritizedJobs) {
+        this.jobsForNextRun.unshift(job);
+      }
+      this.buffer = null;
+      setTimeout(this.runJob, 0);
+    }
+  };
+
   private runJob = async (): Promise<void> => {
     if (this.closed) {
       return;
     }
+    // Prepare work package
+    const prepared = this.prepareWork();
+    if (prepared.length === 0) {
+      return;
+    }
+
+    if (this.blsPoolType === BlsPoolType.workers) {
+      await this.runJobWorkerPool(prepared);
+    } else {
+      await this.runJobLibuv(prepared);
```

Comment:
if possible I'd like to refactor the current implementation to have `BaseBlsMultiThreadPool` to move all common code there, have `BlsMultiThreadWorkerPool` for worker specific implementation, and have this new work in another implementation, so there would be multiple steps required to make this happens similar to `n_historical_states` work I've worked on

that would make a smooth transition to new bls implementation, it's also easier to review and maintain. What's your opinions on this approach @matthewkeil @wemeetagain ?

Reply:
Sounds good.  I can take a look at this in the updated PR #6362 once the metrics look good.  Going to try and limit the complexity until I can get it running smoothly though @tuyennhv